cmd/terminal.go

```diff
@@ -0,0 +1,367 @@
+		case <-ticker.C:
+			pongWaitString := strconv.Itoa(int(pongWait.Seconds()))
+			_ = conn.WriteControl(
+				websocket.PingMessage,
+				[]byte(pongWaitString),
+				time.Now().Add(writeWait),
+			)
+		case <-quit:
+			c.running = false
+		case <-c.stop:
+			c.running = false
+		}
+	}
+
+	// cancel the context
+	cancelContext()
+
+	// send the stop shell message
+	m = &ws.ProtoMsg{
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeShell,
+			MsgType:   wsshell.MessageTypeStopShell,
+			SessionID: c.sessionID,
+		},
+	}
+
+	data, err = msgpack.Marshal(m)
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse the message from the websocket")
+	}
+	_ = conn.SetWriteDeadline(time.Now().Add(writeWait))
+	_ = conn.WriteMessage(websocket.BinaryMessage, data)
+	conn.Close()
+
+	return c.err
+}
+
+func (c *TerminalCmd) resizeTerminal(ctx context.Context, msgChan chan *ws.ProtoMsg, termID int, termWidth int, termHeight int) {
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			newTermWidth, newTermHeight, _ := terminal.GetSize(termID)
+			if newTermWidth != termWidth || newTermHeight != termHeight {
+				termWidth = newTermWidth
+				termHeight = newTermHeight
+				m := &ws.ProtoMsg{
+					Header: ws.ProtoHdr{
+						Proto:   ws.ProtoTypeShell,
+						MsgType: wsshell.MessageTypeResizeShell,
+						Properties: map[string]interface{}{
+							"terminal_width":  termWidth,
+							"terminal_height": termHeight,
+						},
+					},
+				}
+				msgChan <- m
+			}
+		}
+	}
+}
+
+func (c *TerminalCmd) Stop() {
+	c.stop <- true
+	c.running = false
+}
+
+func (c *TerminalCmd) pipeStdout(msgChan chan *ws.ProtoMsg, r io.Reader) {
+	s := bufio.NewReader(r)
+	for c.running {
+		raw := make([]byte, 1024)
+		n, err := s.Read(raw)
+		if err != nil {
+			if c.running {
+				log.Err(fmt.Sprintf("error: %v", err))
+			}
+			break
+		}
+		// CLTR+] or EOF, terminate the shell
+		if raw[0] == 29 || raw[0] == 4 {
+			c.Stop()
+			return
+		}
+
+		m := &ws.ProtoMsg{
+			Header: ws.ProtoHdr{
+				Proto:     ws.ProtoTypeShell,
+				MsgType:   wsshell.MessageTypeShellCommand,
+				SessionID: c.sessionID,
+			},
+			Body: raw[:n],
+		}
+		msgChan <- m
+	}
+}
+
+func (c *TerminalCmd) pipeStdin(conn *websocket.Conn, w io.Writer) {
+	for c.running {
+		_, data, err := conn.ReadMessage()
+		if err != nil {
+			if c.running {
+				log.Err(fmt.Sprintf("error: %v", err))
+			}
+			break
+		}
+
+		m := &ws.ProtoMsg{}
+		err = msgpack.Unmarshal(data, m)
+		if err != nil {
+			log.Err(fmt.Sprintf("error: %v", err))
+			break
+		}
+		if m.Header.Proto == ws.ProtoTypeShell && m.Header.MsgType == wsshell.MessageTypeShellCommand {
+			if _, err := w.Write(m.Body); err != nil {
+				break
+			}
+		} else if m.Header.Proto == ws.ProtoTypeShell && m.Header.MsgType == wsshell.MessageTypeSpawnShell {
+			status := m.Header.Properties["status"].(int64)
```

Comment:
This will crash if `status` is not an int64. Should use `status, ok := ...` construct and error checking.